### PR TITLE
docs: example cluster-wide health endpoint

### LINKED ISSUE #14871: Document CCNP for the health endpoint
Despite https://github.com/cilium/cilium/commit/839030549275119406c9f9f859a3bd90afa1d138, the health endpoint cannot be selected in policies with an `endpointSelector`:
```
$ cat health.yaml 
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "cilium-health-checks"
spec:
  endpointSelector:
    matchLabels:
      reserved: health
  ingress:
    - fromEntities:
      - world
$ cilium policy get
[
  {
    "endpointSelector": {
      "matchLabels": {
        "any:reserved": "health"
      }
    },
    "ingress": [
      {
        "fromEntities": [
          "world"
        ]
      }
    ],
    "labels": [
      {
        "key": "io.cilium.k8s.policy.derived-from",
        "value": "CiliumClusterwideNetworkPolicy",
        "source": "k8s"
      },
      {
        "key": "io.cilium.k8s.policy.name",
        "value": "cilium-health-checks",
        "source": "k8s"
      },
      {
        "key": "io.cilium.k8s.policy.uid",
        "value": "f19c1a4d-f0b2-458d-aeb2-c014273368eb",
        "source": "k8s"
      }
    ]
  }
]
Revision: 2
$ cilium endpoint list | grep health
343        Disabled           Disabled          4          reserved:health                                   fd00::9636   10.11.0.209   ready
$ sudo cilium bpf policy get 343
POLICY   DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS   
Allow    Ingress     reserved:unknown              ANY          NONE         0       0         
Allow    Ingress     reserved:host                 ANY          NONE         8604    93        
Allow    Ingress     reserved:remote-node          ANY          NONE         0       0         
Allow    Egress      reserved:unknown              ANY          NONE         7248    81
```

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <Shikugawa@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #14871 

This PR adds some descriptions and examples to configure cluster-wide health endpoint.

### CODE REVIEW

File: Documentation/policy/intro.rst
Comment (qmonnet) [inline Documentation/policy/intro.rst]: Nitpick: Period at the end of the sentence. Could you also please wrap the line on ~80 characters? Although this document is not really consistent on this, most of the documentation does it.

```suggestion
If you want to configure health entity to check cluster-wide connectivity when
you start cilium-agent with ``enable-policy=always``, see :ref:`health_endpoint`.
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution! This looks good to me on the documentation side, please just find a nitpick below.

Also, sorry to ask, but did you use your real name in the `Signed-off-by:` tag? Please note that [this cannot be a pseudonym](https://docs.cilium.io/en/latest/contributing/development/contributing_guide/#developer-s-certificate-of-origin).

**Comment by Shikugawa** [comment]
> Also, sorry to ask, but did you use your real name in the `Signed-off-by:` tag? Please note that [this cannot be a pseudonym](https://docs.cilium.io/en/latest/contributing/development/contributing_guide/#developer-s-certificate-of-origin).

Ah, sorry. I will fix it.

File: Documentation/policy/intro.rst
Comment (pchaigno) [reply Documentation/policy/intro.rst]: I think we should move this note right after the `always` paragraph on line 33. Otherwise, users may miss it.

Also, the cluster will not be reported as healthy without this policy, so maybe we could reword this as something like: `With that mode you will likely want to enable communications to and from the health endpoint. See ...`. WDYT?

File: examples/policies/kubernetes/clusterwide/health.yaml
Comment (pchaigno) [inline examples/policies/kubernetes/clusterwide/health.yaml]: Did you test this policy with `enable-policy=always`? I don't think it's enough to allow health checks. It would likely need an egress rule as well and maybe `remote-node` or `health` instead of `world`.

**Review by pchaigno — CHANGES_REQUESTED** [review]
Thanks for the pull request! :rocket: 

I don't think the policy for the health endpoint is correct. It would probably be worth testing if you haven't already.

File: examples/policies/kubernetes/clusterwide/health.yaml
Comment (Shikugawa) [reply examples/policies/kubernetes/clusterwide/health.yaml]: With this change, it converted from (A) to (B). Isn't it enough test?

(A)
```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ sudo cilium bpf policy get 299
POLICY   DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS   
Allow    Ingress     reserved:host                 ANY          NONE         7381    80        
Allow    Ingress     reserved:remote-node          ANY          NONE         5384    62   
```

(B)
```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ sudo cilium bpf policy get 299
POLICY   DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS   
Allow    Ingress     reserved:host                 ANY          NONE         7381    80        
Allow    Ingress     reserved:health               ANY          NONE         0       0         
Allow    Ingress     reserved:remote-node          ANY          NONE         5384    62        
Allow    Egress      reserved:health               ANY          NONE         0       0
```

File: Documentation/policy/intro.rst
Comment (Shikugawa) [reply Documentation/policy/intro.rst]: I agree with this idea. Fixed them.

File: examples/policies/kubernetes/clusterwide/health.yaml
Comment (pchaigno) [reply examples/policies/kubernetes/clusterwide/health.yaml]: So that shows that the BPF policy map contains what the K8s policy dictates, but I don't think what the K8s policy dictates is what we actually want.

On your cluster with that policy and `enable-policy=always`, is `cilium-health status --probe` reporting a healthy status? Happy to discuss on Slack if you need help to define the minimum required policy.

File: Documentation/policy/intro.rst
Comment (qmonnet) [inline Documentation/policy/intro.rst]: It looks like you copy-pasted Paul's suggestion, please make sure the resulting sentence reads well.

```suggestion
  If you want to configure health entity to check cluster-wide connectivity when 
  you start cilium-agent with ``enable-policy=always``, you will likely want to
  enable communications to and from the health endpoint.
  See :ref:`health_endpoint`.
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
Thanks for fixing the message and the SOB tag!

One more remark on the new version, below.

Please also note that your changes belong to a single logical commit. This is not an issue to have several commits with your incremental changes for the review, but you'll need to squash them and to force-push on your branch at some point before we merge the PR.

File: examples/policies/kubernetes/clusterwide/health.yaml
Comment (pchaigno) [inline examples/policies/kubernetes/clusterwide/health.yaml]: I did some tests with a GKE cluster and I don't think this is correct. If I setup the cluster in `enable-policy=always` and load this policy, I get the following packet drops for the health traffic:
```
$ kubectl -n kube-system exec cilium-725nz -- cilium monitor -t policy-verdict | grep "action deny"
Policy verdict log: flow 0x0 local EP ID 12, remote ID remote-node, proto 1, ingress, action deny, match none, 10.164.0.44 -> 10.76.1.11 EchoRequest
Policy verdict log: flow 0x0 local EP ID 12, remote ID remote-node, proto 6, ingress, action deny, match none, 10.164.0.44:52576 -> 10.76.1.11:4240 tcp SYN
Policy verdict log: flow 0x0 local EP ID 12, remote ID remote-node, proto 6, ingress, action deny, match none, 10.164.0.44:52576 -> 10.76.1.11:4240 tcp SYN
Policy verdict log: flow 0x0 local EP ID 12, remote ID remote-node, proto 6, ingress, action deny, match none, 10.164.0.44:52576 -> 10.76.1.11:4240 tcp SYN
Policy verdict log: flow 0x0 local EP ID 12, remote ID remote-node, proto 6, ingress, action deny, match none, 10.164.0.44:52576 -> 10.76.1.11:4240 tcp SYN
Policy verdict log: flow 0x0 local EP ID 12, remote ID remote-node, proto 6, ingress, action deny, match none, 10.164.0.44:52576 -> 10.76.1.11:4240 tcp SYN
[...]
```
<br>

I think I know why you missed this :-)  According to [your previous comment](https://github.com/cilium/cilium/pull/15348#discussion_r594584445), you are using the development VMs (:+1:). Unfortunately, the development VM doesn't use exactly the same Cilium options as the Helm installation (more common way to install for real clusters). In particular, the development VM uses `enable-remote-node-identity=false` which leads to the automatic installation of the following rule in each endpoint:
```
POLICY   DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS       
Allow    Ingress     reserved:remote-node          ANY          NONE         5384    62 
```
Once we set `enable-remote-node-identity=true` (default in the GKE deployment for instance), that rule is removed and health check packets are dropped. So I think we should write the policy to work with the most common scenario (`enable-remote-node-identity=true`). Here is the policy I had to use to allow communication to the health endpoints:
```yaml
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "health"
spec:
  endpointSelector:
    matchLabels:
      'reserved:health': ''
  ingress:
  - fromEntities:
    - remote-node
```

I've filed https://github.com/cilium/cilium/issues/15364, a new `good-first-issue`, to change the development VMs and prevent this from happening again in the future.

Comment (Shikugawa) [reply examples/policies/kubernetes/clusterwide/health.yaml]: Yes. I used development VM to conduct experiment. Thank you for pointing out!

Comment (Shikugawa) [reply examples/policies/kubernetes/clusterwide/health.yaml]: So to speak, should we have `remote-node` and `health` entity for healthcheck ingress/egress? like below.
```yaml
spec:
  endpointSelector:
    matchLabels:
      'reserved:health': ''
  ingress:
  - fromEntities:
    - health
    - remote-node
  egress:
  - toEntities:
    - health
    - remote-node
```

Comment (pchaigno) [reply examples/policies/kubernetes/clusterwide/health.yaml]: I think only `remote-node` is needed. I didn't see any health-to-health traffic in my cluster.

**Review by pchaigno — APPROVED** [review]
Could you squash the commits into a single one? We should be good to merge after that.

**Review by qmonnet — APPROVED** [review]
Thanks a lot!

**Comment by pchaigno** [comment]
I don't think we need to wait for the @cilium/policy review as I tested this. Marking as ready to merge.

**Pull request merged**
